src/impl_/extract_argument.rs

```diff
@@ -38,7 +38,7 @@ where
 pub fn extract_argument_with_default<'py, T>(
     obj: Option<&'py PyAny>,
     arg_name: &str,
-    default: impl FnOnce() -> T,
+    default: fn() -> T,
```

Comment:
This implies that argument extraction includes dynamic dispatch? Are we confident on the performance implications of this? Or does it actually improve performance by improving code density?

Reply:
Great question, yes I was also concerned about this and thanks for encouraging me to write down my conclusions!

The primary motivation for this is that it reduces the quantity of llvm code for these methods twofold:
- fewer monomorphizations by removing the generic over `default` (note still generic over `T`)
- I saw the generated code itself is smaller, I would guess because calling `fn() -> T` is simpler than going through the `FnOnce` trait call shim

I believe that LLVM is still happy to inline through `fn()` calls where it knows the callsite - e.g. see https://godbolt.org/z/rYxToExc6. It looks like in that snippet that `example::main` has inlined the call to `do_thing ` and additionally the "dynamic" call to `get_x` so that ultimately it just pushes `5` straight onto the register without doing any function calls at all.

So the conclusion I've reached is that because these functions are all generic (thus code available for inlining) there's a good chance that LLVM is going to inline away the call to `default` completely (and maybe even the call to `extract_argument_with_default`).

Hopefully the outcome of this change is that we're passing a smaller blob of stuff to LLVM, saving on compile times / giving it less to optimize away, while still ending up with an equivalent result.

I looked at our existing benchmarks and I couldn't see any meaningful differences in performance. I could write a pathological case which exercises a lot of defaults, if you'd like to be extra sure.

Reply:
Thank you for the thorough write up. So we get equivalent performance at shorter compile times. Nice!

Reply:
That's my understanding. If we do get users reporting performance concerns we can always revisit in the future. I think the thing to optimize next for performance is still the thread local used in `GILPool` - i.e. I really hope to look at #1308 again, probably in 0.18 when pyproto complexity is gone.